docs/internal/interop.md

```diff
@@ -0,0 +1,351 @@
+   module, the language parser will convert the signature of its own AST into
+   Verona AST and add that node in its own module, then pass that to the Verona
+   AST node.
+
+This allows lazy evaluation of what could be very large number of includes, with
+code that can't be translated into Verona.
+
+The end result is an AST that has foreign nodes as Verona nodes on Verona modules
+without the actual implementation. For all purposes, this is a plain Verona AST
+with calls to functions that have been declared, but not defined.
+
+Definition (and implementation in object code) will be done later by the compiler
+once we know all functions that have survived the initial round of cleanups (ex.
+reachability analysis).
+
+### Compiler
+
+1. For each sandbox region, the compiler generates calls to the runtime to setup
+   the region, including its allocators.
+2. Upon finding a foreign call, the compiler will ask the language driver for
+   the foreign AST node, to implement the actual call.
+3. The driver will generate the conversion between Verona and native types,
+   including structure layout handling per architecture, serialise the
+   arguments and return values into a buffer and call the dispatcher below.
+4. A dispatcher will be created to call those functions with serialised
+   arguments by an index that is associated when they're created.
+5. A bi-directional RPC layer is created between the Verona code and the sandbox,
+   where calls to external functions go one way, and calls to allocation, system
+   calls, etc. go the other way.
+6. The appropriate object files are created and linked with the main executable.
+
+Not all functions will be representable in Verona, not all structure layouts will
+be easily known at the Verona level (special hardware ABI), so errors may still
+occur here.
+
+The compiler is responsible for emitting errors that inform the programmer what
+type/function cannot be represented and potentially why. The programmer is
+responsible for implementing a wrapper function that abstracts away the
+complexity of the language in a public function that Verona can call.
+
+Language support will be always evolving. This is expected to be a balance
+between "best effort" and "appropriate support".
+
+### Run time
+
+1. When the sandbox is initialised, its allocator sets up a new isolated heap
+   that is used for all allocations in the sandbox (including Verona foreign
+   objects that belong to that region).
+2. When the foreign code makes calls to allocate memory or run potentially unsafe
+   operations (file descriptors, system calls, etc.), the RPC layer will check
+   them and forward if explicitly marked as an upcall or if it's a direct
+   sandbox functionality (webasm, CHERI), otherwise fail and close the sandbox.
+3. When Verona code makes calls to the foreign library, the compiler-generated
+   serialisation code will transform the arguments and return value into a
```

Comment:
I'd prefer marshalling to serialisation here.  We're not really serialising, we're just putting the values in a shared buffer.